## BEGIN ##

Copyright 2003 Tim Yamin <plasmaroo@gentoo.org>
                         <plasmaroo@plasmaroo.squirrelserver.co.uk>
All rights reserved.

This file may freely be distributed under the conditions of the 
GPLv2 as long as this header, and the following disclaimer is preserved.

THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY 
EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE 
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR 
PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE FREEBSD PROJECT OR 
CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, 
EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, 
PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR 
PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF 
LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING 
NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS 
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

## END ##
diff -Naur linux-2.6.0/include/linux/blockgroup_lock.h linux-2.6.0/include/linux.plasmaroo/blockgroup_lock.h
--- linux-2.6.0/include/linux/blockgroup_lock.h	2003-12-26 22:19:45.000000000 +0000
+++ linux-2.6.0/include/linux.plasmaroo/blockgroup_lock.h	2003-12-26 19:37:37.000000000 +0000
@@ -5,7 +5,7 @@
  */
 
 #include <linux/config.h>
-#include <linux/spinlock.h>
+#include <linux/spinlock_wrapper.h>
 #include <linux/cache.h>
 
 #ifdef CONFIG_SMP
diff -Naur linux-2.6.0/include/linux/compiler-gcc2.h linux-2.6.0/include/linux.plasmaroo/compiler-gcc2.h
--- linux-2.6.0/include/linux/compiler-gcc2.h	2003-12-26 22:19:44.000000000 +0000
+++ linux-2.6.0/include/linux.plasmaroo/compiler-gcc2.h	2003-12-26 19:37:36.000000000 +0000
@@ -12,7 +12,9 @@
 # define __builtin_expect(x, expected_value) (x)
 #endif
 
+#ifndef __attribute_used__
 #define __attribute_used__	__attribute__((__unused__))
+#endif
 
 /*
  * The attribute `pure' is not implemented in GCC versions earlier
diff -Naur linux-2.6.0/include/linux/ext2_fs_sb.h linux-2.6.0/include/linux.plasmaroo/ext2_fs_sb.h
--- linux-2.6.0/include/linux/ext2_fs_sb.h	2003-12-26 22:19:46.000000000 +0000
+++ linux-2.6.0/include/linux.plasmaroo/ext2_fs_sb.h	2003-12-26 19:37:38.000000000 +0000
@@ -18,6 +18,7 @@
 
 #include <linux/blockgroup_lock.h>
 #include <linux/percpu_counter.h>
+#include <linux/redefine_types.h>
 
 /*
  * second extended-fs super-block data in memory
diff -Naur linux-2.6.0/include/linux/fb.h linux-2.6.0/include/linux.plasmaroo/fb.h
--- linux-2.6.0/include/linux/fb.h	2003-12-26 22:19:46.000000000 +0000
+++ linux-2.6.0/include/linux.plasmaroo/fb.h	2003-12-26 19:37:38.000000000 +0000
@@ -2,9 +2,7 @@
 #define _LINUX_FB_H
 
 #include <linux/tty.h>
-#include <linux/workqueue.h>
 #include <asm/types.h>
-#include <asm/io.h>
 
 /* Definitions of frame buffers						*/
 
@@ -331,6 +329,8 @@
 #define FB_PIXMAP_IO      4     /* memory is iomapped       */
 #define FB_PIXMAP_SYNC    256   /* set if GPU can DMA       */
 
+#ifdef __KERNEL__
+
 struct fb_pixmap {
         __u8  *addr;                      /* pointer to memory             */  
 	__u32 size;                       /* size of buffer in bytes       */
@@ -344,7 +344,9 @@
 	spinlock_t lock;                  /* spinlock                      */
 	atomic_t count;
 };
-#ifdef __KERNEL__
+
+#include <linux/workqueue.h>
+#include <asm/io.h>
 
 #include <linux/fs.h>
 #include <linux/init.h>
diff -Naur linux-2.6.0/include/linux/fs.h linux-2.6.0/include/linux.plasmaroo/fs.h
--- linux-2.6.0/include/linux/fs.h	2003-12-26 22:19:45.000000000 +0000
+++ linux-2.6.0/include/linux.plasmaroo/fs.h	2003-12-26 19:37:37.000000000 +0000
@@ -17,7 +17,15 @@
 #include <linux/dcache.h>
 #include <linux/stat.h>
 #include <linux/cache.h>
-#include <linux/radix-tree.h>
+
+/* This should fix any ``sys-fs'' items getting spinlocks.
+   $Header: /var/cvsroot/gentoo-x86/sys-kernel/linux-headers/files/linux-headers-2.6.0-appCompat.patch,v 1.7 2003/12/26 23:28:56 plasmaroo Exp $ 
+*/
+
+#ifdef __KERNEL__
+	#include <linux/radix-tree.h>
+#endif
+
 #include <linux/kobject.h>
 #include <asm/atomic.h>
 
diff -Naur linux-2.6.0/include/linux/isdn.h linux-2.6.0/include/linux.plasmaroo/isdn.h
--- linux-2.6.0/include/linux/isdn.h	2003-12-26 22:19:46.000000000 +0000
+++ linux-2.6.0/include/linux.plasmaroo/isdn.h	2003-12-26 19:37:37.000000000 +0000
@@ -14,7 +14,13 @@
 #define __ISDN_H__
 
 #include <linux/ioctl.h>
-#include <linux/isdn/fsm.h>
+
+/* This should sort out ``ligtop'' from getting spinlocks.
+   $Header: /var/cvsroot/gentoo-x86/sys-kernel/linux-headers/files/linux-headers-2.6.0-appCompat.patch,v 1.7 2003/12/26 23:28:56 plasmaroo Exp $
+*/
+#ifdef __KERNEL__
+	#include <linux/isdn/fsm.h>
+#endif
 
 #ifdef CONFIG_COBALT_MICRO_SERVER
 /* Save memory */
diff -Naur linux-2.6.0/include/linux/mod_devicetable.h linux-2.6.0/include/linux.plasmaroo/mod_devicetable.h
--- linux-2.6.0/include/linux/mod_devicetable.h	2003-12-26 22:19:46.000000000 +0000
+++ linux-2.6.0/include/linux.plasmaroo/mod_devicetable.h	2003-12-26 19:37:38.000000000 +0000
@@ -9,8 +9,10 @@
 
 #ifdef __KERNEL__
 #include <linux/types.h>
-typedef unsigned long kernel_ulong_t;
+#else
+#include <linux/redefine_types.h>
 #endif
+typedef unsigned long kernel_ulong_t;
 
 #define PCI_ANY_ID (~0)
 
diff -Naur linux-2.6.0/include/linux/msdos_fs.h linux-2.6.0/include/linux.plasmaroo/msdos_fs.h
--- linux-2.6.0/include/linux/msdos_fs.h	2003-12-26 22:19:46.000000000 +0000
+++ linux-2.6.0/include/linux.plasmaroo/msdos_fs.h	2003-12-26 19:37:37.000000000 +0000
@@ -4,7 +4,14 @@
 /*
  * The MS-DOS filesystem constants/structures
  */
-#include <linux/buffer_head.h>
+
+/* This should sort out any problems with ``dosfstools''.
+   $Header: /var/cvsroot/gentoo-x86/sys-kernel/linux-headers/files/linux-headers-2.6.0-appCompat.patch,v 1.7 2003/12/26 23:28:56 plasmaroo Exp $
+*/
+#ifdef __KERNEL__
+	#include <linux/buffer_head.h>
+#endif
+
 #include <linux/string.h>
 #include <asm/byteorder.h>
 
diff -Naur linux-2.6.0/include/linux/percpu_counter.h linux-2.6.0/include/linux.plasmaroo/percpu_counter.h
--- linux-2.6.0/include/linux/percpu_counter.h	2003-12-26 22:19:46.000000000 +0000
+++ linux-2.6.0/include/linux.plasmaroo/percpu_counter.h	2003-12-26 19:37:37.000000000 +0000
@@ -5,10 +5,13 @@
  */
 
 #include <linux/config.h>
-#include <linux/spinlock.h>
-#include <linux/smp.h>
+#include <linux/spinlock_wrapper.h>
 #include <linux/threads.h>
 
+/* <smp.h> was removed as it brought in spinlocks and other
+   bad things - no functionality is lost as all the necessary
+   SMP data is still available. */
+
 #ifdef CONFIG_SMP
 
 struct __percpu_counter {
diff -Naur linux-2.6.0/include/linux/redefine_types.h linux-2.6.0/include/linux.plasmaroo/redefine_types.h
--- linux-2.6.0/include/linux/redefine_types.h	1970-01-01 01:00:00.000000000 +0100
+++ linux-2.6.0/include/linux.plasmaroo/redefine_types.h	2003-12-26 19:37:38.000000000 +0000
@@ -0,0 +1,46 @@
+/* This is a short macro to redefine any types which weren't included
+   further down the tree as __KERNEL__ might have been unset. 
+
+   $>  find $path -name types.h -maxdepth 2 | xargs cat | grep \#ifndef | grep \
+       TYPES | sed 's/ifndef/undef/'
+
+   plasmaroo@gentoo.org - This is an "as-is" Public Domain Item with 
+                          no warranty of any kind */
+
+#ifndef __KERNEL__
+	#define __KERNEL__
+	#define __KERNEL_WAS_DEFINED__
+#endif
+
+#undef __ASSEMBLY__ 
+#undef _ALPHA_TYPES_H
+#undef __ASM_ARM_TYPES_H
+#undef _ASM_TYPES_H
+#undef __UM_TYPES_H
+#undef _LINUX_TYPES_H
+#undef __BIT_TYPES_DEFINED__
+#undef _ASM_IA64_TYPES_H
+#undef _S390_TYPES_H
+#undef _PPC_TYPES_H
+#undef __ASM_SH_TYPES_H
+#undef _SPARC64_TYPES_H
+#undef _ASM_TYPES_H
+#undef _SPARC_TYPES_H
+#undef _LINUX_RXRPC_TYPES_H
+#undef _PARISC_TYPES_H
+#undef __ASM_ARM_TYPES_H
+#undef _X86_64_TYPES_H
+#undef _ALPHA_TYPES_H
+#undef _M68K_TYPES_H
+#undef __V850_TYPES_H__
+#undef _ETRAX_TYPES_H
+#undef _M68K_TYPES_H
+#undef _I386_TYPES_H
+#undef _H8300_TYPES_H
+#undef _PPC64_TYPES_H
+
+#include <asm/types.h>
+
+#ifdef __KERNEL_WAS_DEFINED__
+	#undef __KERNEL__
+#endif
diff -Naur linux-2.6.0/include/linux/spinlock_wrapper.h linux-2.6.0/include/linux.plasmaroo/spinlock_wrapper.h
--- linux-2.6.0/include/linux/spinlock_wrapper.h	1970-01-01 01:00:00.000000000 +0100
+++ linux-2.6.0/include/linux.plasmaroo/spinlock_wrapper.h	2003-12-26 19:37:36.000000000 +0000
@@ -0,0 +1,26 @@
+#ifndef __LINUX_SPINLOCK_H_WRAPPER
+#define __LINUX_SPINLOCK_H_WRAPPER
+
+#include <linux/preempt.h>
+
+#define SPINLOCK_MAGIC  0x1D244B3C
+typedef struct {
+	unsigned long magic;
+	volatile unsigned long lock;
+	volatile unsigned int babble;
+	const char *module;
+	char *owner;
+	int oline;
+} spinlock_t;
+
+#define spin_lock_init(x) \
+	do { \
+		(x)->magic = SPINLOCK_MAGIC; \
+		(x)->lock = 0; \
+		(x)->babble = 5; \
+		(x)->module = __FILE__; \
+		(x)->owner = NULL; \
+		(x)->oline = 0; \
+	} while (0)
+
+#endif
diff -Naur linux-2.6.0/include/linux/time.h linux-2.6.0/include/linux.plasmaroo/time.h
--- linux-2.6.0/include/linux/time.h	2003-12-26 22:19:46.000000000 +0000
+++ linux-2.6.0/include/linux.plasmaroo/time.h	2003-12-26 19:37:38.000000000 +0000
@@ -4,24 +4,31 @@
 #include <asm/param.h>
 #include <linux/types.h>
 
+#ifndef _SYS_TIME_H
+
 #ifndef _STRUCT_TIMESPEC
+#ifndef __timespec_defined
 #define _STRUCT_TIMESPEC
 struct timespec {
 	time_t	tv_sec;		/* seconds */
 	long	tv_nsec;	/* nanoseconds */
 };
+#endif /* __timespec_defined */
 #endif /* _STRUCT_TIMESPEC */
 
+#ifndef _STRUCT_TIMEVAL
 struct timeval {
 	time_t		tv_sec;		/* seconds */
 	suseconds_t	tv_usec;	/* microseconds */
 };
+#endif /* _STRUCT_TIMEVAL */
 
 struct timezone {
 	int	tz_minuteswest;	/* minutes west of Greenwich */
 	int	tz_dsttime;	/* type of dst correction */
 };
 
+#endif /* _SYS_TIME_H */
 #ifdef __KERNEL__
 
 #include <linux/spinlock.h>
@@ -351,16 +358,23 @@
 #define	ITIMER_VIRTUAL	1
 #define	ITIMER_PROF	2
 
+#ifndef _TIME_H
+
 struct  itimerspec {
         struct  timespec it_interval;    /* timer period */
         struct  timespec it_value;       /* timer expiration */
 };
 
+#endif
+
+#ifndef _SYS_TIME_H
+
 struct	itimerval {
 	struct	timeval it_interval;	/* timer interval */
 	struct	timeval it_value;	/* current value */
 };
 
+#endif /* _SYS_TIME_H */
 
 /*
  * The IDs of the various system clocks (for POSIX.1b interval timers).
diff -Naur linux-2.6.0/include/linux/videodev.h linux-2.6.0/include/linux.plasmaroo/videodev.h
--- linux-2.6.0/include/linux/videodev.h	2003-12-26 22:19:46.000000000 +0000
+++ linux-2.6.0/include/linux.plasmaroo/videodev.h	2003-12-26 19:37:37.000000000 +0000
@@ -3,7 +3,13 @@
 
 #include <linux/types.h>
 #include <linux/version.h>
-#include <linux/device.h>
+
+/* Should sort out ``DirectFB''.
+   $Header: /var/cvsroot/gentoo-x86/sys-kernel/linux-headers/files/linux-headers-2.6.0-appCompat.patch,v 1.7 2003/12/26 23:28:56 plasmaroo Exp $
+*/
+#ifdef __KERNEL__
+	#include <linux/device.h>
+#endif
 
 #define HAVE_V4L2 1
 #include <linux/videodev2.h>
